### PR TITLE
[test] Add infrastructure for integration testing with Docker containers

### PR DESCRIPTION
**Pull request by jaylees14**
- Add infrastructure to call `Docker` containers from within integration tests
- Added a basic integration test to show how the tools work -- this should probably be replaced with real tests, although could be good to keep around to make sure we don't break anything

Currently, **unit tests** are ran in SBT with

``` 
sbt:temple> test
```

**Integration tests** are ran with 

```
sbt:temple> it:test
```

I have not enabled integration testing via the CI - should we do this?

### CODE REVIEW

File: src/it/scala/temple/containers/DockerPostgresService.scala [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  val databaseUsername = "postgres"
+  val databasePassword = ""
+  val databaseName     = "temple_test"
+  val internalPort     = 5432
+  val externalPort     = 44444
+  val externalUrl      = s"jdbc:postgresql://localhost:$externalPort/$databaseName"
+
+  class PostgresReadyChecker extends DockerReadyChecker {
+
+    // Called when container has started
```
Comment (jaylees14) [inline src/it/scala/temple/containers/DockerPostgresService.scala]: This comment is wrong. Should be “called every time, but only succeeds when started”

File: build.sbt [hunk 2]
```diff
@@ -4,5 +4,17 @@ version := "0.1"
 
 scalaVersion := "2.13.1"
 
-libraryDependencies += "org.scala-lang.modules" %% "scala-parser-combinators" % "1.1.2"
-libraryDependencies += "org.scalatest" %% "scalatest" % "3.0.8" % "test"
+lazy val root = (project in file("."))
```
Comment (c-harding) [inline build.sbt:8]: plz cite, this is mad

File: src/it/scala/temple/containers/DockerPostgresService.scala [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        val url = s"jdbc:postgresql://${docker.host}:$externalPort/"
+        Option(DriverManager.getConnection(url, databaseUsername, databasePassword))
+          .foreach { conn =>
+            try conn.createStatement().execute(s"CREATE DATABASE $databaseName")
+            finally conn.close()
+          }
+        Future.successful(true)
+      } catch {
+        case error: Throwable => Future.failed(error)
+      }
```
Comment (c-harding) [inline src/it/scala/temple/containers/DockerPostgresService.scala]:
```scala
val url = s"jdbc:postgresql://${docker.host}:$externalPort/"
Future {
  Option(DriverManager.getConnection(url, databaseUsername, databasePassword)) map { conn =>
    try conn.createStatement().execute(s"CREATE DATABASE $databaseName")
    finally conn.close()
  }
}

File: src/it/scala/temple/containers/DockerPostgresService.scala [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  }
+}
+
+/** DockerPostgresService configures a docker container for Postgres */
+trait DockerPostgresService extends DockerKit {
+
+  val postgresContainer: DockerContainer = DockerContainer(DockerPostgresService.image)
+    .withPorts((DockerPostgresService.internalPort, Some(DockerPostgresService.externalPort)))
```
Comment (c-harding) [inline src/it/scala/temple/containers/DockerPostgresService.scala]:
```scala
withPorts(DockerPostgresService.internalPort -> Some(DockerPostgresService.externalPort))
```
not sure, thoughts?

File: src/it/scala/temple/containers/PostgresSpec.scala [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import com.spotify.docker.client.DefaultDockerClient
+import com.whisk.docker.DockerFactory
+import com.whisk.docker.impl.spotify.SpotifyDockerFactory
+import com.whisk.docker.scalatest.DockerTestKit
+import org.scalatest.{BeforeAndAfterAll, FlatSpec, Matchers}
+
+/** PostgresSpec offers additional methods to test commands using Postgres database */
+trait PostgresSpec extends FlatSpec with DockerTestKit with DockerPostgresService with BeforeAndAfterAll {
+
+  implicit override val dockerFactory: DockerFactory = new SpotifyDockerFactory(DefaultDockerClient.fromEnv().build())
```
Comment (c-harding) [inline src/it/scala/temple/containers/PostgresSpec.scala:15]: wtf is this Java stuff?

File: src/it/scala/temple/containers/PostgresSpec.scala [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    Option(
+      DriverManager.getConnection(
+        DockerPostgresService.externalUrl,
+        DockerPostgresService.databaseUsername,
+        DockerPostgresService.databasePassword
+      )
+    ).map { conn =>
+      try fn(conn)
+      finally conn.close()
+    }
```
Comment (c-harding) [inline src/it/scala/temple/containers/PostgresSpec.scala]:
```scala
Option {
  DriverManager.getConnection(
    DockerPostgresService.externalUrl,
    DockerPostgresService.databaseUsername,
    DockerPostgresService.databasePassword
  )
} map { conn =>
  try fn(conn)
  finally conn.close()
}
```
?

**Review by c-harding — COMMENTED** [review]
Some random scalafications, feel free to ignore if you want

File: src/it/scala/temple/containers/DockerPostgresService.scala (hunk 3 above)
Comment (jaylees14) [reply src/it/scala/temple/containers/DockerPostgresService.scala]: Not sure this works - the type this returns is `Future[Unit]` not the required `Future[Boolean]`

Comment (c-harding) [reply src/it/scala/temple/containers/DockerPostgresService.scala]: Add a line only containing true just before the Future closes

File: src/it/scala/temple/containers/PostgresSpec.scala (hunk 6 above)
Comment (c-harding) [reply src/it/scala/temple/containers/PostgresSpec.scala]: Not sure whether this is preferable, but the Driver Manager feels like too much of a code block to fit within parens. There’s no hard-and-fast rule of when to use which though

**Review by c-harding — APPROVED** [review]

**Review by Lewiky — APPROVED** [review]
LGTM

**Pull request merged**